### ISSUE TITLE
Confusing error reported by the validator caused by compiler failing to report an incorrect use of function parameter 'out' qualifier

### ISSUE THREAD

**Issue by jshofmann** [microsoft/DirectXShaderCompiler#6398]
**Description**
When a function parameter used as an inout parameter is erroneously labeled as just an out parameter, dxc fails to report the error but the DXIL validator does catch it. This results in a couple of very confusing error messages reported by dxc. If validation is disabled via -Vd then dxc will produce a buggy shader without error or warning.

**Steps to Reproduce**
See the attached file 'UninitializedValue.hlsl.txt' containing a minimal repro case. Strip off the '.txt' extension github demanded and compile the file with 'dxc -T ps_6_2 -E PixelShader UninitializedValue.hlsl'.

**Actual Behavior**
You should see these errors from the validator:
UninitializedValue.hlsl:21:23: error: Instructions should not read uninitialized value.
    note: at '%16 = fmul fast float %13, undef' in block '#0' of function 'PixelShader'.
UninitializedValue.hlsl:21:39: error: Instructions should not read uninitialized value.
    note: at '%22 = fmul fast float %19, undef' in block '#0' of function 'PixelShader'.

To fix the error, in the function _HueControls replace the 'out float3 hue' parameter with 'inout float3 hue'. An error referencing the problematic HLSL instead of the generated DXIL would save a lot of time identifying the badly specified parameter and allow our technical art staff to fix the problem without requiring support from the engineering team.

**Environment**
- DXC version: dxcompiler.dll: 1.7 - 1.7.2308.7 (69e54e290); dxil.dll: 1.7(101.7.2308.12)
- Host Operating System: Windows 11 Pro 23H2 build 22631.3235 with Windows Feature Experience Pack 1000.22687.1000.0
[UninitializedValue.hlsl.txt](https://github.com/microsoft/DirectXShaderCompiler/files/14531264/UninitializedValue.hlsl.txt)

**Comment by llvm-beanz** [comment]
[Compiler Explorer](https://godbolt.org/z/do1Kcajbd)

**Comment by llvm-beanz** [comment]
I had previously done some work to improve these diagnostics. The original PR was #5047, but due to inaccuracies in our AST representation for some data types we needed to limit the types it applied to which we did in #5078.

The core problem is that we don't represent class and struct types (which include vectors and matrices) accruately. I had a draft PR with a partial fix for the bigger AST problem posted as #5249. I don't suspect we're likely to land that PR ever due to the huge scale and massive risk (it also isn't complete). Our real solution for this issue will be to do it better in Clang rather than to fix all the bugs in DXC.

I realize this isn't a satisfying answer, but due to the complexity of this problem and our plans to move away from DXC in favor of Clang we're just not going to fix this.
